src/xml2_node.cpp

```diff
@@ -27,13 +27,57 @@ std::string nodeName(T* node, SEXP nsMap) {
   return prefix + ":" + name;
 }
 
+SEXP node_name_impl(SEXP x, SEXP nsMap) {
+  NodeType type = getNodeType(x);
+
+  SEXP out;
+
+  switch(type) {
+  case NodeType::missing:
+    out = NA_STRING;
+    break;
+  case NodeType::node: {
+    SEXP node_sxp = VECTOR_ELT(x, 0);
+    XPtrNode node(node_sxp);
+
+    std::string name = nodeName(node.checked_get(), nsMap);
+    out = Rf_mkCharLenCE(name.c_str(), name.size(), CE_UTF8);
+    break;
+  }
+  default: Rf_error("Unexpected node type");
```

Comment:
Maybe extract out a `stop_unexpected_node_type()` helper that you can call, in case we ever want to tweak the message?